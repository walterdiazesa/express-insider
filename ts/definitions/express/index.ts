export enum HandlerType {
  ROUTE = "Route",
  MIDDLEWARE = "Middleware",
  ROUTE_MIDDLEWARE = "Route Middleware",
}

export * from './objects';
export * from './verbs';