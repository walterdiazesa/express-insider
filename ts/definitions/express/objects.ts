import { TrailOptions } from "..";
import { HandlerType, Method } from "../../../ts";

export type RouteMatcher = {
  route: TrailOptions['ignoreRoutes'][number]['route'],
  method: Lowercase<TrailOptions['ignoreRoutes'][number]['method']>,
};

export type Route = {
  path: `/${string}`;
  stack: RouteStack[];
  methods: Record<Method, boolean>;
};

type Stack<T extends HandlerType | undefined = undefined> = {
  handle: Function & { stack?: (StackItem | StackItem<HandlerType.ROUTE>)[] };
  name: string;
  params?: undefined;
  path: T extends HandlerType.ROUTE ? string : unknown;
  keys: Record<string, string | boolean | number>[];
  regexp: RegExp;
};

export type StackItem<T extends HandlerType | undefined = undefined> = Stack<T> & {
  route: T extends undefined ? undefined | Route : T extends HandlerType.ROUTE ? Route : undefined;
  // Config params
  ignore: T extends HandlerType.ROUTE ? undefined | true : never;
  showResponse: T extends HandlerType.ROUTE ? undefined | true : never;
  showRequestedURL: T extends HandlerType.ROUTE ? undefined | true : never;
}

export type RouteStack<T extends Method = Method> = Stack & {
  method: T;
};