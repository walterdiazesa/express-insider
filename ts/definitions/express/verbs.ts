export type Method = "get" | "post" | "options" | "all" | "patch" | "delete" | "put" | "head" | "connect" | "trace";