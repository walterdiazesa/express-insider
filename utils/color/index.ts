import { COLOR } from "../../constants";

export const getStatusCodeColor = (statusCode: number) => (statusCode >= 200 && statusCode <= 299 ? COLOR.fgGreen : statusCode >= 400 ? COLOR.fgRed : COLOR.fgWhite);