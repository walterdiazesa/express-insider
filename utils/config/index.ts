import { Find, Method, Route, RouteMatcher, TrailOptions } from "../../ts";
import { Config, ConfigKeys } from "../../ts/definitions/app/config";

// const digest = (property: LooseAutocomplete<ConfigKeys[number]>): keyof ConfigMapper => {
export const digest = <T extends ConfigKeys[number]>(property: T): Find<ConfigKeys, T> => {
  switch (property) {
    // @ts-ignore
    case "trailId": return 0;
    // @ts-ignore
    case "logger": return 1;
    // @ts-ignore
    case "ignoreMiddlewares": return 2;
    // @ts-ignore
    case "ignoreRoutes": return 3;
    // @ts-ignore
    case "showRequestedURL": return 4;
    // @ts-ignore
    case "showResponse": return 5;
    // @ts-ignore
    case "showRSS": return 6;
    // @ts-ignore
    case "showColors": return 7;
    // @ts-ignore
    case "report": return 8;
    // @ts-ignore
    case "timingFormatter": return 9;
    // @ts-ignore
    case "initialImmutableMiddlewares": return 10;
    // @ts-ignore
    case "logStrategy": return 11;
    // @ts-ignore
    case "delayMs": return 12;
    // @ts-ignore
    case "skip": return 13;
    /* istanbul ignore next */
    default:
      // TypeScript will raise an error if the default case is reached.
      const _: never = property;
      throw new Error(`Property "${String(property)}" not supported.`)
  }
}

export const generateRouteMatcherGroup = (matcher: Config['ignoreRoutes'] | boolean) => {
  if (typeof matcher === 'boolean') return matcher;
  const matchGroup: (`/${string}` | RouteMatcher['method'][])[] = [] as any;
  for (let i = 0; i < matcher.length; i++) {
    const matchItem = matcher[i];
    const enlisted = matchGroup.findIndex((route) => route === matchItem.route);
    const method = matchItem.method.toLowerCase() as RouteMatcher['method'];
    if (enlisted !== -1) {
      const enlistedItem = matchGroup[enlisted + 1] as RouteMatcher['method'][];
      if (!enlistedItem.includes(method)) enlistedItem.push(method)
    }
    else matchGroup.push(matchItem.route, [method])
  }
  return matchGroup;
}