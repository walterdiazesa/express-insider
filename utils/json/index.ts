import { COLOR, TYPE_TO_COLOR, JSONTagPad } from "../../constants";

const valueFormatter = (value: any) => {
  const formatted =  typeof value === "string" ? `${COLOR.fgGreen}"${value}"` : `${TYPE_TO_COLOR[value === false || value === null ? "red" : typeof value]}${value}`
  return `${COLOR.reset}${formatted}${COLOR.reset}`
};
export const colorizedJSON = (nextLinePad: string, value: any, objMultiline = true, firstMultiline = true): string => {
  if (typeof value !== "object" || value === null) return valueFormatter(value)
  if (Array.isArray(value)) return `[\n${nextLinePad}${JSONTagPad}${value
      .map(item => colorizedJSON(`${nextLinePad}${JSONTagPad}`, item, objMultiline)).join(`,\n${nextLinePad}${JSONTagPad}`)}\n${nextLinePad}]`;
  const objEntries = Object.entries(value);
  if (!objEntries.length || !objEntries.some(([key, value]) => (typeof value === "object" ? Object.keys(value).length : true))) return "";
  if (!firstMultiline) return `{ ${objEntries.filter(([qK, qV]) => qV !== undefined).map(([qK, qV]) => `${COLOR.fgYellow}"${qK}": ${colorizedJSON(nextLinePad, qV, objMultiline)}`).join(", ")} }`
  return `{${objEntries
    .reduce((acc, [key, value]) => {
      if (value === undefined) return acc;
      const elements = `${acc},\n${nextLinePad}${JSONTagPad}${COLOR.fgYellow}${key}: ${COLOR.reset}`
      if (typeof value !== "object" || value === null) return `${elements}${valueFormatter(value)}`;
      const valueKeys = Object.keys(value);
      if (!valueKeys.length) return acc;
      return `${elements}${colorizedJSON(`${nextLinePad}${JSONTagPad}`, value, objMultiline, objMultiline)}`;
    }, "")
    .slice(1)}\n${nextLinePad}}`;
};