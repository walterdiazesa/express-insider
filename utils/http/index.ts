import { Response } from "express";
import { HTTP_STATUS_MESSAGE_TO_CODE } from "../../constants";

/**
  * __[ODD-1]__: When using express@4.9.0 (unknown about other versions under 4.10.0) the test "trail on custom configuration â€º Should handle [CASE 12] with POST /conditional-route"
  * fails, even that the route return statusCode 200 as expected, the trail output shows 404, when removing the next call in the third route everything works as expected,
  * what is happenning is that express treats undefined next routes different in 4.9.0, so as no afterwards middleware is found the app crashes setting the statusCode as 404, so
  * the bandage fix to support 4.9.0 is to redefine the possible outcome for 404 cases if res.statusMessage is still defined as "OK"
  * 
  * Current state: FIXED!, we could return => res.statusCode directly, but it's a good redundant condition
  */
export const getStatusCode = (res: Response): typeof HTTP_STATUS_MESSAGE_TO_CODE[keyof typeof HTTP_STATUS_MESSAGE_TO_CODE] =>
  res.statusCode === 404 && res.statusMessage !== 'Not found' ? HTTP_STATUS_MESSAGE_TO_CODE[res.statusMessage.toLowerCase().replaceAll(' ', '')] : res.statusCode;