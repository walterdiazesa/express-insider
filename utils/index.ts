import { getCfg } from '../config'
import { Express } from 'express'
import { HandlerType, StackItem, StackItemType } from '../ts';

export const getRSS = (): '' | ` [${number} bytes]` => {
  if (!getCfg().showRSS) return ''
  // BUN: memoryUsage on process since v0.6.14
  let rss = process.memoryUsage?.().rss;
  if (rss === undefined) {
    try {
      // BUN: bun:jsc available since v0.5.1
      const { memoryUsage } = require('bun:jsc');
      rss = memoryUsage().current
    } catch (e) {}
  }
  return ` [${rss} bytes]`
}
export const getStackItemType = (stackItem: StackItem): StackItemType => (stackItem.route ? HandlerType.ROUTE : HandlerType.MIDDLEWARE);

export const formatAnonymousRoute = (idx: number): `<anonymous (${number})>` => `<anonymous (${idx})>`;

/**
 * __[ODD-3]__: Under 4.0 ```app.stack``` is defined, after is ```app._router.stack```
 */
export const getStack = (app: Express): StackItem[] => app.stack || app._router.stack

export * from './report'
export * from './json'
export * from './color'
export * from './logger'
export * from './http'