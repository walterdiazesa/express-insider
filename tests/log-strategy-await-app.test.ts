import request from 'supertest'
import express from 'express'
import { trail } from '../index';
import { awaitConsoleLog } from './utils';

const sleep = (ms: number, payload?: object) => new Promise((r) => setTimeout(() => r(payload), ms));

const appAwaitEach = express();

appAwaitEach.get('/', async (req, res) => {
  await sleep(100);
  res.send({ working: true })
})

appAwaitEach.get('/fail-route', (req, res) => {
  res.status(404).send({ working: false })
})

trail(appAwaitEach, {
  trailId: () => 'test-id',
  timingFormatter: (elapsed) => 0,
  logStrategy: 'await-each',
  skip: (req, res) => res.statusCode === 404,
});

describe('trail on logStrategy await-each configuration', () => {
  it('Should handle GET /', async () => {
    const log = jest.spyOn(global.console, 'log');
    request(appAwaitEach).get("/").then();
    expect(log.mock.calls.length).toBe(0);
    await sleep(300)
    expect(log.mock.calls).toMatchSnapshot();
  })
  it('Should ignore GET /fail-route', async () => {
    const log = jest.spyOn(global.console, 'log');
    await request(appAwaitEach).get("/fail-route");
    await awaitConsoleLog(0, log.mock.calls);
    expect(log.mock.calls.length).toBe(0);
  })
});