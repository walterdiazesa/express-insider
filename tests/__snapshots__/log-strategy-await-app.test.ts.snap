// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`trail on logStrategy await each configuration Should handle GET / 1`] = `
[
  [
    "[test-id] [32mGET / [0m[1mstart[0m
[test-id] [36mMiddleware [32mquery [33m0 ms[0m
[test-id] [36mMiddleware [32mexpressInit [33m0 ms[0m
[test-id] [35mRoute [32mGET / [33m0 ms [32m200 OK[0m
[test-id] [32mGET / [0m[1mfinish[0m, elapsed time since begin: [33m0 ms[0m",
  ],
]
`;
